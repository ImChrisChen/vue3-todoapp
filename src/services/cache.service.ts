interface CacheOptions {
  time?: number
  id?: string
}

interface CacheContent extends CacheOptions {
  value: unknown
}

// 需要使用Key时需要在这里注册
type CacheKey = 'todolist'

class CacheService {
  cacheKey = 'TodoAppCache:'
  defaultExpireTime = 1000 * 60 * 15 // 默认缓存时间15分钟

  private getKey(key: string) {
    return this.cacheKey + key
  }

  setItem(key: CacheKey, value, options: CacheOptions = {}) {
    try {
      const finalKey = this.getKey(key)
      const content: CacheContent = {
        time: (options.time || this.defaultExpireTime) + Date.now(),
        id: options.id || '',
        value: value,
      }
      localStorage.setItem(finalKey, JSON.stringify(content))
      return true
    } catch (e) {
      return false
    }
  }

  getItem(key: CacheKey, options: Omit<CacheOptions, 'time'> = {}) {
    if (!key) {
      new Error('key为必须')
      return null
    }

    let content = {} as CacheContent

    try {
      const finalKey = this.getKey(key)
      const jsonStr: string = localStorage.getItem(finalKey) || '{}'
      content = JSON.parse(jsonStr)
    } catch (e) {
      console.error(e)
    }

    // id mismatch
    if (options.id && content.id && options.id !== content.id) {
      return null
    }

    // availability time
    if (!content.time || content.time < Date.now()) {
      return null
    }

    return content.value
  }

  removeItem(key: string) {
    return localStorage.removeItem(key)
  }
}

export const cacheService = new CacheService()
