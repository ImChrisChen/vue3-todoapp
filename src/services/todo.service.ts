import httpClient, { ResponseData } from '@/http/http'

export declare namespace TodoDataDto {
  interface CreateTodoDto {
    name: string
    done?: boolean
  }

  interface UpdateTodoDto {
    name?: string
    done?: boolean
  }
}

class TodoService {
  async update<T>(id: number, data: TodoDataDto.UpdateTodoDto) {
    return httpClient.put(`/todo/${id}`, data)
  }

  async delete(id: number) {
    return httpClient.delete(`/todo/${id}`)
  }

  async getTodoList<T>(): Promise<T | any[]> {
    return httpClient.get<ResponseData<T>>(`/todo/`).then((res) => {
      if (res.data.result !== 'ok') {
        console.error('')
        // return res.data.data
      }
      return res?.data?.data || []
    })
  }

  async createTodo(data: TodoDataDto.CreateTodoDto) {
    if (!('done' in data)) {
      data.done = false
    }
    if (!data.name) {
      // console.error('name不能为空', data)
      throw new Error(`name不能为空`)
    }

    return httpClient.post<ResponseData<any>>(`/todo`, data).then((res) => {
      return res.data
    })
  }
}

export const todoService = new TodoService()
