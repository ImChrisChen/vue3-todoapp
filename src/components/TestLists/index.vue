<template>
  <ul>
    <li :class="{ li: true }" v-for="(item, i) in items" :key="i">
      {{ item.name }}
    </li>
  </ul>
</template>

<script lang="ts" setup name="TestLists">
import { Todo, todoService } from '@/services'
import { onMounted, ref } from 'vue'

const items = ref<Todo[]>([])

onMounted(() => {
  todoService.getTodoList().then((list) => {
    items.value = list
  })
})
</script>

<style scoped lang="less">
.done {
  color: gray;
}
</style>
