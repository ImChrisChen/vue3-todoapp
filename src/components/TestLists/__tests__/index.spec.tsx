import { mount, VueWrapper } from '@vue/test-utils'
import TestLists from '../index.vue'
import { todoService } from '@/services'
import { h } from 'vue'
const todolist = require('@mock/todo-list.json')

todoService.getTodoList = jest.fn().mockResolvedValue(todolist)

describe('components/TestLists', function () {
  const testLists: VueWrapper = mount(<TestLists />)

  it('should render items', async () => {
    const len = testLists.findAll('.li')
    expect(len).toHaveLength(todolist.length)
    expect(testLists).toMatchSnapshot() // 没有快照则会在当前目录生成一个快照(__snapshot__)
  })

  // it('TestLists Snapshot Test', () => {
  //   expect(testLists.element).toMatchSnapshot()
  // })
})
