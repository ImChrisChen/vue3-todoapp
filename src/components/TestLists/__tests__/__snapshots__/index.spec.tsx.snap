// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TestLists TestLists Snapshot Test 1`] = `
"<ul>
  <li class=\\"li\\">洗澡12313</li>
  <li class=\\"li\\">起床</li>
  <li class=\\"li\\">11点洗澡</li>
  <li class=\\"li\\">提交代码</li>
</ul>"
`;

exports[`components/TestLists should render items 1`] = `
"<ul>
  <li class=\\"li\\">洗澡12313</li>
  <li class=\\"li\\">起床</li>
  <li class=\\"li\\">11点洗澡</li>
  <li class=\\"li\\">提交代码</li>
</ul>"
`;
